homeassistant/components/denonavr/__init__.py

```diff
@@ -33,3 +48,70 @@ def service_handler(service):
         hass.services.register(DOMAIN, service, service_handler, schema=schema)
 
     return True
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up the denonavr components from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Connect to receiver
+    connect_denonavr = ConnectDenonAVR(
+        hass,
+        entry.data[CONF_HOST],
+        DEFAULT_TIMEOUT,
+        entry.options.get(CONF_SHOW_ALL_SOURCES, DEFAULT_SHOW_SOURCES),
+        entry.options.get(CONF_ZONE2, DEFAULT_ZONE2),
+        entry.options.get(CONF_ZONE3, DEFAULT_ZONE3),
+    )
+    if not await connect_denonavr.async_connect_receiver():
+        raise ConfigEntryNotReady
+    receiver = connect_denonavr.receiver
+
+    hass.data[DOMAIN][entry.entry_id] = receiver
+
+    entry.add_update_listener(update_listener)
+
+    device_registry = await dr.async_get_registry(hass)
+    if entry.data[CONF_MAC] is not None:
+        device_registry.async_get_or_create(
```

Comment:
You could save a few lines with something like this

```
       device_info = {
           "identifiers": ...
       }
       if entry.data[CONF_MAC] is not None:
           device_info["connections"] = {(dr.CONNECTION_NETWORK_MAC, entry.data[CONF_MAC])}

        device_registry.async_get_or_create(**device_info)
```

Reply:
Actually it probably makes this a lot simpler if all of this is moved into `def device_info` in `media_player.py`

Reply:
If I move it to the device itself, I would have to pass all the config_entry data required for it, so I would prefer to keep it in the __init__.py file

Reply:
From my read of the documentation, it implies that `async_get_or_create` is for when there is no entity so I think it should be moved into `device_info`

We could wait for Martin to give a second opinion though if you don't concur.

Reply:
Lets wait on the revieuw of @MartinHjelmare to see what he thinks. I don't have strong feelings on it. I'm fine with moving it to the entity, but personly i think it is cleaner as it is now. So I will let @MartinHjelmare decide...

Reply:
We should only call `device_registry.async_get_or_create` when we need to create a device that is not the same device as the device that has a feature represented by an entity. Eg if there's a hub that connects multiple other devices, we can create a home assistant device to represent the hub, while the other devices will be represented by entities which will automatically create home assistant devices via the `device_info` property.

Reply:
I will move this to the entity then, will work on it this evening

Reply:
Done and tested